File: src/leetcode/python/BinarySearchTreeIterator_173.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BinarySearchTreeIterator_173(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.cache = []
        while root:
            self.cache.append(root)
            root = root.left

    def hasNext(self):
        """
        :rtype: bool
        """
        return len(self.cache) != 0

    def next(self):
        """
        :rtype: int
        """
        node = self.cache.pop()
        res = node.val
        node = node.right
        while node:
            self.cache.append(node)
            node = node.left
        return res

File: src/leetcode/java/BinarySearchTreeIterator_173.java
package leetcode.java;

import java.util.LinkedList;

public class BinarySearchTreeIterator_173 {
  private LinkedList<TreeNode> cache;
  
  public BinarySearchTreeIterator_173(TreeNode root) {
    this.cache = new LinkedList<>();
    while (root != null) {
      cache.add(root);
      root = root.left;
    }
  }

  /** @return whether we have a next smallest number */
  public boolean hasNext() {
    return cache.size() != 0;
  }

  /** @return the next smallest number */
  public int next() {
    TreeNode node = cache.removeLast();
    int res = node.val;
    node = node.right;
    while (node != null) {
      cache.add(node);
      node = node.left;
    }
    return res;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
